### ISSUE TITLE
[OJS] Updating to OJS3 - revision files and initial copyedit files missing

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#2097]
Hi,

It seems that revision files and initial copyedit files are missing after upgrading to OJS3. The files do exist but the database is not getting upgraded correctly.

More details here:
http://forum.pkp.sfu.ca/t/updating-from-ojs2-to-ojs-3-all-revisions-and-initial-copyedits-missing/26631

**Comment by ajnyga** [comment]
Hi @bozana and @asmecher 

I did a clean install of 2.4.8.1. added on article and the upgraded to OJS3. A few things I found out.

There is a line in `review_round_files`. However, the value in the `file_id` field is wrong. The value given there is the file_id for the original review file. And while that file has file stage set to 4, it does not appear in the revisions. In my test case just by chancing the value in the `file_id` field in `review_round_files` made the revision file appear.

However, this introduced another problem. In OJS2 I added two revision files as an author. It seems that the migration process still writes only one line to `review_round_files` and even when I fix the file_id problem I can only see one file. 

Many authors tend to send multiple files at this stage and they are not the revision of the same file. One of the files could be a report of the changes they made to the manuscript. So after the migration, the revisions panel should show all the same files as in OJS2 in the same view.

**Comment by ajnyga** [comment]
Regarding the other issue I mentions, showing more that one revision file, you have to edit `controllers/grid/files/SubmissionFilesGridDataProvider.inc.php` => `prepareSubmissionFileData` line 164 or you have to skip `prepareSubmissionFileData` in `controllers/grid/files/review/ReviewGridDataProvider.inc.php`.

**Comment by asmecher** [comment]
@ajnyga, would you be willing to provide a database dump to work with? (Anonymized as you like, and sent privately, of course.)

**Comment by ajnyga** [comment]
Hi,

I will anonymize names, emails, article titles and abstracts and send it to you. Our database is around 500 mb. Do you have suggestions on how to send that? Of course this same problem appears also in the clean install I made and that is a small database with only a few users, one journal and article.

**Comment by bozana** [comment]
I took a look at the revisions files upgrade problem: They are not considered/inserted into the table `review_round_files`. This can be easily done for the upgrade from 2.4.x to 3.0.x (with the SQL from @ajnyga above), but it will be a little bit difficult to repair it for those that already upgraded and are using 3.0.x, because some information is missing -- we do not know the review round any more (from `article_files`). @asmecher, do you have an idea what to do in that case? Would it be acceptable to insert all revision files in the current/latest or in the first review round?
@asmecher, what would you suggest for the display of those revision files? The revisions in OJS 2.4.x are treated as revisions from the same file -- the file id is the same and revision changes. But this seems to be different than the users expect i.e. how the revisions are displayed in the OJS 3.0.x. I am not sure we can and if it makes sense to display a revision per row here, because it is not how it functions everywhere else in the system and for the future revisions? Maybe to somehow provide the access to the earlier revisions in that grid, but I am not sure if this is satisfying for the users -- if they expect them to be independent files and see them as such?

**Comment by ajnyga** [comment]
Hi,

Thanks @bozana

I think that we have two somewhat separate issues. Firstly the issue of migrating revisions to OJS3 and second, the way the revisions are shown in OJS3.

I know that there are a lot of situations where an author submits more than one file when sending revisions and usually it is not the case the she would be sending several revisions of the same file. Usually the second file can be for example a report concerning the changes the author has made to the manuscript.

On database level OJS2 seems to think these are the revisions of the same file. But in the UI this is not clear at all as you can see from this screenshot with two revisions:
![screenshot_22](https://cloud.githubusercontent.com/assets/16347527/21353704/e5929b2c-c6cf-11e6-8c96-464561900b91.png)

Now that this is upgraded to OJS3, the second file is simply lost and only the pdf (5-2) is visible and while the author has intended to send two separate files, this is of course not expected behaviour.

The fix a told above changes the revision view so that all the versions are listed (there was mistake there which I corrected) like this:
![screenshot_24](https://cloud.githubusercontent.com/assets/16347527/21353823/68615908-c6d0-11e6-9857-eca86399fa77.png)

So this solves the scenario I mentioned above and at least at the moment I can not figure out what the problem would be? On the contrary I would like to see as an editor all the revisions of the same file the author has sent to a specific review round and you can still say what is the latest file there (marked with (2)).

**Comment by bozana** [comment]
Hmmm... If I see it correctly the copyediting files are migrated to the copyediting discussions. @asmecher, is this OK/correct/meant to be so? 
They are also revisions of the same file i.e. with the same file id and different revision numbers. They are displayed as all revisions i.e. only the latest.

**Comment by asmecher** [comment]
Hi @bozana and @ajnyga:

From @bozana:
> I took a look at the revisions files upgrade problem: They are not considered/inserted into the table review_round_files. This can be easily done for the upgrade from 2.4.x to 3.0.x (with the SQL from @ajnyga above), but it will be a little bit difficult to repair it for those that already upgraded and are using 3.0.x, because some information is missing -- we do not know the review round any more (from article_files). @asmecher, do you have an idea what to do in that case? Would it be acceptable to insert all revision files in the current/latest or in the first review round?
I would prefer not to make assumptions about missing data -- if the assumption is wrong and people start reporting that file access is not as expected, then we'll have to consider four scenarios instead of the current two. Better to correct the upgrade scripts for new upgrades and leave already-performed upgrades as they are, though perhaps with some documentation here about resolutions for users who need to correct their data. This is not optimal but the 2.x to 3.x upgrade process is complex and some of these things will happen, unfortunately.

> @asmecher, what would you suggest for the display of those revision files? The revisions in OJS 2.4.x are treated as revisions from the same file -- the file id is the same and revision changes. But this seems to be different than the users expect i.e. how the revisions are displayed in the OJS 3.0.x. I am not sure we can and if it makes sense to display a revision per row here, because it is not how it functions everywhere else in the system and for the future revisions? Maybe to somehow provide the access to the earlier revisions in that grid, but I am not sure if this is satisfying for the users -- if they expect them to be independent files and see them as such?
If I understand the problem correctly, it's that reviewers have been using the revision system to store different files (e.g. that aren't revisions). That's not expected behavior and I'm not sure we should support it. Fortunately OJS 3 now considers pools of files, not single files, so the work-around isn't needed for future reviews/submissions.

**Comment by ajnyga** [comment]
Hi @asmecher and @bozana 

Great to hear that you are taking this into account in the upgrade script. We are upgrading to OJS3 in mid January and together with #2047 and #1782 this is the most important thing to get fixed.

Alec you probably meant to write that "authors have been using the revisions system to store different files" - not reviewers. That is the case. 

Basically the authors have used the " Upload Author Version" function to upload not only a revision of the actual manuscript but also revisions of other files and also new files altogether, like reports on the changes they have done etc. I realize that the authors should have used the "Add a supplementary file" function from the summary page, but OJS2 has not been especially clear in this regard and uploading the other files together with the revision has been an easy way for the authors to do that. OJS2 does not prevent it in any way and only now I too realized that these files also have a common fileId.

So although this is not exptected behaviour it is probably fairly common behaviour in other installations as well and at least in our case we have to support the journals who have these cases. Otherwise they will basically loose data which is especially problematic in submission undergoing editorial process during the upgrade.

I have three questions here:
1. Do the changes I mention above break something? (I did not see any problems)
2. How hard would it be to get the other revision files linked in the discussion panel as a new discussion? This could be a work-around here?
3. @asmecher what is your opinion on the filenames? I suggested that the migrated files should be named with the original filename because I think it would clarify things (like 1-2-2-ED.docx)

**Comment by bozana** [comment]
PRs
ojs master: https://github.com/pkp/ojs/pull/1159
@asmecher, could you take a look at the PR?

**Comment by bozana** [comment]
Concerning copyediting:
If there is only one initial copyediting file, it will have the file stage 9 i.e. it should be presented in the copyedited grid. However it is not presented there, because of `viewable = 0`, which is also the problem here: https://github.com/pkp/pkp-lib/issues/2092. I.e. those viewable issues still have to be corrected.

If there were more copyediting rounds, the files will have file stage 18 and will be presented in the copyediting discussions grid. Here however, as for revisions, only the last revision will be displayed. Also the naming is same as for revisions i.e. the name will be from the first revision, leading to: if the first revision was a PDF and the last revision was a DOC, the name will be with PDF but the downloaded file will be DOC :-)

**Comment by asmecher** [comment]
@ajnyga, there are many parts of OJS2 where only the latest revision is available via the workflow, and you'd have to go back to the Files Manager to access previous revisions. We can do better than that with OJS3 -- I agree that we should provide links to previous revisions -- but haven't gotten to it yet. (Filed: https://github.com/pkp/pkp-lib/issues/2130)

I haven't heard of anyone else using the author uploads this way. I'd suggest making a change to the upgrade process to split these out into separate files, rather than continuing to track them as revisions.

Naming files per the OJS2 filename would be relatively simple -- as long as this means setting the `original_file_name` in the database, not trying to maintain the OJS2-style filename on the server filesystem.

**Comment by asmecher** [comment]
@bozana, pkp/ojs#1159 looks good to me, thanks!

**Comment by ajnyga** [comment]
Thanks for the comments and changes!

The problem with splitting the files is that I do not know which files to split and checking it manually is not really an option. Now that you have #2130 I will probably apply the fix I suggested above to get the revisions to show and when #2130 is ready, I will remove my own fix. The thing is that I am upgrading our server to OJS3 next Monday and need to make sure that those files are readable for the editors right after the upgrade so that they can continue their work.

As for the file names, I was definitely thinking only about the name stored in the database which is shown to the users - not the actual file name in the filesystem. However, I was thinking about changing the `name` values in `submission_file_settings`, not the `original_file_name` in `submission_files`. 

But I can see now that the database structure in OJS3 does not allow me to name the revisions separately in `submission_file_settings`. So what I am planning to do is to rename the files using the `file_name` for revision 1 from `article_files_migration`.

[Side note: The fact that the file name shows the original file format is a bit confusing if the file format changes in a revision. The original file name is still visible in the name if you do not rename it manually during upload]

**Comment by bozana** [comment]
The one initial copyedited file that had `viewable = 0` and that was migrated to the Copyedited grid, but wasn't visible, should be visible in that grid now i.e. with the second PR from here: https://github.com/pkp/pkp-lib/issues/2092#issuecomment-270604156.

**Comment by ajnyga** [comment]
Hi @bozana,

Thank you so much for the work you have done.
I can confirm that with these changes to the upgrade script I was able to migrate our 45 journals with all revisions visible. Also the fix in #2092 seems to be working as expected!

The only problem with many submissions is that in the journals have not uploaded a Layout version in OJS2 meaning that after migrating to OJS3 the links to published articles go to the copyeding page in the dashboard. Or at least I think that it is caused by that. The status of the submission in the link label is correct. But this is a small problem and can probably be fixed fairly easily later if needed.

For the problem with multiple revision files I noticed that the solution I mentioned above does break some things, so I had to make some corrections there. I just need a solution for the next couple of months that the journal editors can access all the files the authors have sent them, I will follow https://github.com/pkp/pkp-lib/issues/2130 and see what the final solution will be with that issue. 

Regarding the file naming I ended up with doing a change in the upgrade script. Instead of using the naming pattern username, usergroup, original filename, I used the original `file_name` from `article_files_migration` table with small changes. Basically I have renamed for example 2-13-1-ED.docx to "2-13-ED". I also ended up renaming all submission files with the name "Untitled" in a similar manner. Basically 99% of old supplementary files have that name, because the authors do not bother to name them in OJS2.

So basically with these:
![screenshot_45](https://cloud.githubusercontent.com/assets/16347527/21742490/c98cf8f0-d4f8-11e6-8032-0573f0c011e7.png)
Instead of this:
![screenshot_46](https://cloud.githubusercontent.com/assets/16347527/21742492/cd7096a2-d4f8-11e6-87c2-a6786eab6c15.png)
I get this:
![screenshot_47](https://cloud.githubusercontent.com/assets/16347527/21742493/d0ea067e-d4f8-11e6-9c73-0d5bf8d6ec6e.png)

**Comment by bozana** [comment]
@ajnyga, in this issue only the question of file names is left, right?

**Comment by ajnyga** [comment]
Yes I think so. The issue with the revisions is handled in #2130 and for me, the solution there is enough at this time. 

Regarding the file names I guess it is mostly a matter of opinion. I noticed that you can actually modify it from the locale files. I ended up doing the name conversion shown above, but the new files are getting the default name format.

I still think that sticking with the original name of the file in that file name fields has at least these problems:
1. The file extension is shown in the name and does not change although a revision file is in different format
2. The fact that the original filename is shown there may raise questions among the authors. Think of a situation where author uploads a file called "my-article.docx" to OJS and later comes back to upload a revision named "my-revised-article-final.docx". After uploading that new revision, system is still showing this name the author: "Username, Author, my-article.docx". 

But as I said, this is a matter of opinion. One format I have been thinking of is this: [Username], [Role],[Component][#id], Revision [(#Rev)]. The output would something like: "ajnyga, Author, Article text 2348, Revision (1)"

**Comment by bozana** [comment]
@asmecher, what do you think about those file names i.e. shall we do/change something there, now or later, or...? If we would like to do something maybe to create a new issue for that, to be able to close this one? :-)

**Comment by asmecher** [comment]
We've gotten some UI/UX testing feedback on filenames too -- tagging @stranack about the suggestion above -- but I think we should deal with it outside this entry.

**Comment by bozana** [comment]
I opened a new issue for the file revision names here: https://github.com/pkp/pkp-lib/issues/2235
For another file names issue, see https://github.com/pkp/pkp-lib/issues/2003
Closing this issue...
